File: bot/src/voice/under-music.ts
/**
 * Voice under music + progressive enhancement (V1 / H4).
 * Pure helpers for smoke tests and admin check API — no live TS required.
 */

import { extractWatchwordCommand } from "./watchword.js";

export interface UnderMusicConfig {
  duckMusicOnSpeech: boolean;
  duckMusicVolume: number;
  listenWindowMs: number;
  textWakeFallback: boolean;
  watchword: string;
}

export interface UnderMusicPlan {
  duckActive: boolean;
  duckLevel: number;
  listenWindowMs: number;
  /** How wake is accepted when KWS is unavailable. */
  progressiveWake: "text-fallback" | "kws-only" | "armed-followup";
  textFallbackAlwaysWorks: boolean;
  notes: string[];
}

export function defaultUnderMusicConfig(partial?: Partial<UnderMusicConfig>): UnderMusicConfig {
  return {
    duckMusicOnSpeech: partial?.duckMusicOnSpeech !== false,
    duckMusicVolume: clamp(
      partial?.duckMusicVolume === undefined ||
        partial.duckMusicVolume === 2 ||
        partial.duckMusicVolume === 25
        ? 20
        : partial.duckMusicVolume,
      0,
      100,
    ),
    listenWindowMs: Math.max(15_000, partial?.listenWindowMs ?? 15_000),
    textWakeFallback: partial?.textWakeFallback !== false,
    watchword: (partial?.watchword ?? "moneypenny").toLowerCase(),
  };
}

function clamp(n: number, lo: number, hi: number): number {
  return Math.max(lo, Math.min(hi, n));
}

/** Operator-facing plan: duck + listen window + progressive text wake. */
export function planUnderMusicCapture(cfg: UnderMusicConfig): UnderMusicPlan {
  const notes: string[] = [];
  if (cfg.duckMusicOnSpeech) {
    notes.push(`Music ducks to volume ${cfg.duckMusicVolume} while listening.`);
  } else {
    notes.push("Duck is off — STT under DJ load will struggle.");
  }
  notes.push(`Post-wake listen window ${cfg.listenWindowMs}ms (min 15s).`);
  if (cfg.textWakeFallback) {
    notes.push("Text wake fallback ON — “Moneypenny pause” works without KWS (Whisper path).");
  } else {
    notes.push("Text wake fallback OFF — requires KWS or armed follow-up.");
  }
  notes.push("Typed/chat commands always work as text fallback for transport.");

  return {
    duckActive: cfg.duckMusicOnSpeech,
    duckLevel: cfg.duckMusicVolume,
    listenWindowMs: cfg.listenWindowMs,
    progressiveWake: cfg.textWakeFallback ? "text-fallback" : "kws-only",
    textFallbackAlwaysWorks: true,
    notes,
  };
}

export interface UnderMusicTurnResult {
  transcript: string;
  matched: boolean;
  command: string;
  path: "text-wake" | "kws" | "armed" | "none";
  /** Would music be ducked for this capture path? */
  wouldDuck: boolean;
  /** Safe chat/text equivalent when voice fails. */
  textFallbackCommand: string | null;
}

/**
 * Simulate one under-music voice turn using the real watchword extractor.
 * Progressive: text wake works without KWS; armed follow-ups work after wake.
 */
export function simulateUnderMusicTurn(
  transcript: string,
  cfg: UnderMusicConfig,
  opts?: { kwsDetected?: boolean; armed?: boolean },
): UnderMusicTurnResult {
  const text = transcript.trim();
  const match = extractWatchwordCommand(text, cfg.watchword, {
    textWakeFallback: cfg.textWakeFallback,
    kwsDetected: opts?.kwsDetected,
    armed: opts?.armed,
  });

  let path: UnderMusicTurnResult["path"] = "none";
  if (match.matched) {
    if (opts?.kwsDetected) path = "kws";
    else if (opts?.armed && !cfg.textWakeFallback) path = "armed";
    else if (opts?.armed && !text.toLowerCase().includes(cfg.watchword)) path = "armed";
    else if (cfg.textWakeFallback) path = "text-wake";
    else path = "armed";
  }

  const command = match.matched ? match.command.trim() : "";
  const wouldDuck = cfg.duckMusicOnSpeech && match.matched;
  // Progressive text fallback: bare transport verbs via chat always work.
  const textFallbackCommand =
    command ||
    (opts?.armed ? text : null) ||
    (!match.matched && /^(pause|skip|stop|resume|next|prev)\b/i.test(text) ? text : null);

  return {
    transcript: text,
    matched: match.matched,
    command,
    path,
    wouldDuck,
    textFallbackCommand: textFallbackCommand?.trim() || null,
  };
}

export interface UnderMusicSmokeCase {
  id: string;
  transcript: string;
  kwsDetected?: boolean;
  armed?: boolean;
  expectMatched: boolean;
  expectCommandIncludes?: string;
  expectPath?: UnderMusicTurnResult["path"];
}

/** Default smoke cases for V1 under-music reliability (unit-driven). */
export const UNDER_MUSIC_SMOKE_CASES: UnderMusicSmokeCase[] = [
  {
    id: "text-wake-pause",
    transcript: "Moneypenny pause",
    expectMatched: true,
    expectCommandIncludes: "pause",
    expectPath: "text-wake",
  },
  {
    id: "text-wake-skip",
    transcript: "Moneypenny skip",
    expectMatched: true,
    expectCommandIncludes: "skip",
    expectPath: "text-wake",
  },
  {
    id: "armed-followup",
    transcript: "pause",
    armed: true,
    expectMatched: true,
    expectCommandIncludes: "pause",
    expectPath: "armed",
  },
  {
    id: "banter-no-wake",
    transcript: "I need to pause and think",
    expectMatched: false,
    expectPath: "none",
  },
  {
    id: "kws-path",
    transcript: "skip",
    kwsDetected: true,
    expectMatched: true,
    expectCommandIncludes: "skip",
    expectPath: "kws",
  },
];

export interface UnderMusicSmokeReport {
  ok: boolean;
  plan: UnderMusicPlan;
  results: Array<UnderMusicTurnResult & { id: string; pass: boolean; reason?: string }>;
}

export function runUnderMusicSmoke(
  cfg?: Partial<UnderMusicConfig>,
  cases: UnderMusicSmokeCase[] = UNDER_MUSIC_SMOKE_CASES,
): UnderMusicSmokeReport {
  const full = defaultUnderMusicConfig(cfg);
  const plan = planUnderMusicCapture(full);
  const results = cases.map((c) => {
    // Armed-only cases should not require text wake on the bare verb path.
    const caseCfg =
      c.armed && !c.transcript.toLowerCase().includes(full.watchword)
        ? { ...full, textWakeFallback: false }
        : full;
    const r = simulateUnderMusicTurn(c.transcript, caseCfg, {
      kwsDetected: c.kwsDetected,
      armed: c.armed,
    });
    let pass = r.matched === c.expectMatched;
    let reason: string | undefined;
    if (pass && c.expectCommandIncludes) {
      pass = r.command.toLowerCase().includes(c.expectCommandIncludes.toLowerCase());
      if (!pass) reason = `command "${r.command}" missing "${c.expectCommandIncludes}"`;
    }
    if (pass && c.expectPath) {
      pass = r.path === c.expectPath;
      if (!pass) reason = `path ${r.path} !== ${c.expectPath}`;
    }
    if (!pass && !reason) reason = `matched=${r.matched} expected=${c.expectMatched}`;
    return { id: c.id, ...r, pass, reason };
  });
  return {
    ok: results.every((r) => r.pass),
    plan,
    results,
  };
}

File: crates/mp-voice/src/under_music.rs
// Copyright (c) 2026 Lane Ambrose
// SPDX-License-Identifier: MIT

//! Under-music progressive wake (V1/H4). Pure helpers + admin smoke API.

use mp_config::{effective_duck_volume, VoiceConfig, MIN_LISTEN_WINDOW_MS};

use crate::watchword::{extract_watchword_command, WatchwordOptions};

#[derive(Debug, Clone)]
pub struct UnderMusicConfig {
    pub duck_music_on_speech: bool,
    pub duck_music_volume: u32,
    pub karaoke_mode: bool,
    pub listen_window_ms: u64,
    pub text_wake_fallback: bool,
    pub watchword: String,
}

impl UnderMusicConfig {
    pub fn from_voice(cfg: &VoiceConfig) -> Self {
        Self {
            duck_music_on_speech: cfg.duck_music_on_speech,
            karaoke_mode: cfg.karaoke_mode,
            duck_music_volume: effective_duck_volume(cfg.karaoke_mode, cfg.duck_music_volume),
            listen_window_ms: cfg.listen_window_ms.max(MIN_LISTEN_WINDOW_MS),
            text_wake_fallback: cfg.text_wake_fallback,
            watchword: cfg.watchword.to_ascii_lowercase(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum WakePath {
    TextWake,
    Kws,
    Armed,
    None,
}

impl WakePath {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::TextWake => "text-wake",
            Self::Kws => "kws",
            Self::Armed => "armed",
            Self::None => "none",
        }
    }
}

pub fn plan_under_music(cfg: &UnderMusicConfig) -> serde_json::Value {
    let mut notes = Vec::new();
    if cfg.duck_music_on_speech {
        notes.push(if cfg.karaoke_mode {
            format!(
                "Karaoke ON — music ducks to volume {} while listening.",
                cfg.duck_music_volume
            )
        } else {
            format!(
                "Music ducks to volume {} while listening.",
                cfg.duck_music_volume
            )
        });
    } else {
        notes.push("Duck is off — STT under DJ load will struggle.".into());
    }
    notes.push(format!(
        "Post-wake listen window {}ms (min 15s).",
        cfg.listen_window_ms
    ));
    if cfg.text_wake_fallback {
        notes.push(
            "Text wake fallback ON — “Moneypenny pause” works without KWS (Whisper path).".into(),
        );
    } else {
        notes.push("Text wake fallback OFF — requires KWS or armed follow-up.".into());
    }
    notes.push("Typed/chat commands always work as text fallback for transport.".into());
    serde_json::json!({
        "duckActive": cfg.duck_music_on_speech,
        "duckLevel": cfg.duck_music_volume,
        "listenWindowMs": cfg.listen_window_ms,
        "progressiveWake": if cfg.text_wake_fallback { "text-fallback" } else { "kws-only" },
        "textFallbackAlwaysWorks": true,
        "notes": notes,
    })
}

pub fn simulate_under_music_turn(
    transcript: &str,
    cfg: &UnderMusicConfig,
    kws_detected: bool,
    armed: bool,
) -> (bool, String, WakePath, bool, Option<String>) {
    let text = transcript.trim();
    let match_ = extract_watchword_command(
        text,
        &cfg.watchword,
        WatchwordOptions {
            text_wake_fallback: cfg.text_wake_fallback,
            kws_detected,
            armed,
        },
    );
    let mut path = WakePath::None;
    if match_.matched {
        if kws_detected {
            path = WakePath::Kws;
        } else if armed && !cfg.text_wake_fallback {
            path = WakePath::Armed;
        } else if armed && !text.to_ascii_lowercase().contains(&cfg.watchword) {
            path = WakePath::Armed;
        } else if cfg.text_wake_fallback {
            path = WakePath::TextWake;
        } else {
            path = WakePath::Armed;
        }
    }
    let command = if match_.matched {
        match_.command.trim().to_string()
    } else {
        String::new()
    };
    let would_duck = cfg.duck_music_on_speech && match_.matched;
    let fallback = if !command.is_empty() {
        Some(command.clone())
    } else if armed {
        Some(text.to_string())
    } else if !match_.matched
        && regex_transport(text)
    {
        Some(text.to_string())
    } else {
        None
    };
    (match_.matched, command, path, would_duck, fallback.filter(|s| !s.trim().is_empty()))
}

fn regex_transport(text: &str) -> bool {
    let t = text.trim().to_ascii_lowercase();
    ["pause", "skip", "stop", "resume", "next", "prev"]
        .iter()
        .any(|v| t == *v || t.starts_with(&format!("{v} ")))
}

pub struct SmokeCase {
    pub id: &'static str,
    pub transcript: &'static str,
    pub kws_detected: bool,
    pub armed: bool,
    pub expect_matched: bool,
    pub expect_command_includes: Option<&'static str>,
    pub expect_path: Option<WakePath>,
}

pub fn default_smoke_cases() -> Vec<SmokeCase> {
    vec![
        SmokeCase {
            id: "text-wake-pause",
            transcript: "Moneypenny pause",
            kws_detected: false,
            armed: false,
            expect_matched: true,
            expect_command_includes: Some("pause"),
            expect_path: Some(WakePath::TextWake),
        },
        SmokeCase {
            id: "text-wake-skip",
            transcript: "Moneypenny skip",
            kws_detected: false,
            armed: false,
            expect_matched: true,
            expect_command_includes: Some("skip"),
            expect_path: Some(WakePath::TextWake),
        },
        SmokeCase {
            id: "armed-followup",
            transcript: "pause",
            kws_detected: false,
            armed: true,
            expect_matched: true,
            expect_command_includes: Some("pause"),
            expect_path: Some(WakePath::Armed),
        },
        SmokeCase {
            id: "banter-no-wake",
            transcript: "I need to pause and think",
            kws_detected: false,
            armed: false,
            expect_matched: false,
            expect_command_includes: None,
            expect_path: Some(WakePath::None),
        },
        SmokeCase {
            id: "kws-path",
            transcript: "skip",
            kws_detected: true,
            armed: false,
            expect_matched: true,
            expect_command_includes: Some("skip"),
            expect_path: Some(WakePath::Kws),
        },
    ]
}

pub fn run_under_music_smoke(cfg: &UnderMusicConfig) -> serde_json::Value {
    let plan = plan_under_music(cfg);
    let mut results = Vec::new();
    let mut ok = true;
    for c in default_smoke_cases() {
        let mut case_cfg = cfg.clone();
        if c.armed && !c.transcript.to_ascii_lowercase().contains(&cfg.watchword) {
            case_cfg.text_wake_fallback = false;
        }
        let (matched, command, path, would_duck, fallback) =
            simulate_under_music_turn(c.transcript, &case_cfg, c.kws_detected, c.armed);
        let mut pass = matched == c.expect_matched;
        let mut reason = None;
        if pass {
            if let Some(inc) = c.expect_command_includes {
                if !command.to_ascii_lowercase().contains(&inc.to_ascii_lowercase()) {
                    pass = false;
                    reason = Some(format!("command \"{command}\" missing \"{inc}\""));
                }
            }
        }
        if pass {
            if let Some(ep) = &c.expect_path {
                if path != *ep {
                    pass = false;
                    reason = Some(format!("path {} !== {}", path.as_str(), ep.as_str()));
                }
            }
        }
        if !pass && reason.is_none() {
            reason = Some(format!("matched={matched} expected={}", c.expect_matched));
        }
        if !pass {
            ok = false;
        }
        results.push(serde_json::json!({
            "id": c.id,
            "transcript": c.transcript,
            "matched": matched,
            "command": command,
            "path": path.as_str(),
            "wouldDuck": would_duck,
            "textFallbackCommand": fallback,
            "pass": pass,
            "reason": reason,
        }));
    }
    serde_json::json!({ "ok": ok, "plan": plan, "results": results })
}

#[cfg(test)]
mod tests {
    use super::*;
    use mp_config::VoiceConfig;

    #[test]
    fn smoke_passes_defaults() {
        let cfg = UnderMusicConfig::from_voice(&VoiceConfig::default());
        let report = run_under_music_smoke(&cfg);
        assert_eq!(report["ok"], true, "{report}");
    }
}
